File: java/src/main/java/com/newnancity/newnanmanager/services/IpService.java
package com.newnancity.newnanmanager.services;

import com.newnancity.newnanmanager.exceptions.NewNanManagerException;
import com.newnancity.newnanmanager.generated.api.IpServiceApi;
import com.newnancity.newnanmanager.generated.model.BanIPRequest;
import com.newnancity.newnanmanager.generated.model.IPInfo;
import com.newnancity.newnanmanager.generated.model.IPStatistics;
import com.newnancity.newnanmanager.generated.model.ListIPsResponse;
import com.newnancity.newnanmanager.generated.model.UnbanIPRequest;

/** IP lookup, risk and ban management service. */
public final class IpService extends AbstractService {
    private final IpServiceApi api;

    public IpService(IpServiceApi api) {
        this.api = api;
    }

    public void banIP(BanIPRequest request) throws NewNanManagerException {
        executeVoid(() -> api.banIP(request).execute());
    }

    public void unbanIP(UnbanIPRequest request) throws NewNanManagerException {
        executeVoid(() -> api.unbanIP(request).execute());
    }

    public IPInfo getIPInfo(String ip) throws NewNanManagerException {
        return execute(() -> api.getIPInfo(ip).execute());
    }

    public IPStatistics getIPStatistics() throws NewNanManagerException {
        return execute(() -> api.getIPStatistics().execute());
    }

    public ListIPsResponse listIPs() throws NewNanManagerException {
        return listIPs(null, null, null, null, null);
    }

    public ListIPsResponse listIPs(
            Integer page,
            Integer pageSize,
            Boolean bannedOnly,
            Long minThreatLevel,
            Integer minRiskScore) throws NewNanManagerException {
        return execute(() -> api.listIPs()
                .page(page)
                .pageSize(pageSize)
                .bannedOnly(bannedOnly)
                .minThreatLevel(minThreatLevel)
                .minRiskScore(minRiskScore)
                .execute());
    }

    public ListIPsResponse getBannedIPs(
            Integer page,
            Integer pageSize,
            Boolean bannedOnly,
            Long minThreatLevel,
            Integer minRiskScore) throws NewNanManagerException {
        return execute(() -> api.getBannedIPs()
                .page(page)
                .pageSize(pageSize)
                .bannedOnly(bannedOnly)
                .minThreatLevel(minThreatLevel)
                .minRiskScore(minRiskScore)
                .execute());
    }

    public ListIPsResponse getHighRiskIPs(
            Integer page,
            Integer pageSize,
            Boolean bannedOnly,
            Long minThreatLevel,
            Integer minRiskScore) throws NewNanManagerException {
        return execute(() -> api.getHighRiskIPs()
                .page(page)
                .pageSize(pageSize)
                .bannedOnly(bannedOnly)
                .minThreatLevel(minThreatLevel)
                .minRiskScore(minRiskScore)
                .execute());
    }

    public ListIPsResponse getSuspiciousIPs(
            Integer page,
            Integer pageSize,
            Boolean bannedOnly,
            Long minThreatLevel,
            Integer minRiskScore) throws NewNanManagerException {
        return execute(() -> api.getSuspiciousIPs()
                .page(page)
                .pageSize(pageSize)
                .bannedOnly(bannedOnly)
                .minThreatLevel(minThreatLevel)
                .minRiskScore(minRiskScore)
                .execute());
    }

    public IpServiceApi raw() {
        return api;
    }
}

File: kotlin/src/main/kotlin/com/nanmanager/bukkit/services/IPService.kt
package com.nanmanager.bukkit.services

import com.nanmanager.bukkit.http.HttpClient
import com.nanmanager.bukkit.models.*

/** IP 管理服务；传输和模型解析由 OpenAPI 生成客户端负责。 */
class IPService(httpClient: HttpClient) : BaseService(httpClient) {
    private val api get() = httpClient.generatedApis.ips

    fun getIPInfo(request: GetIPInfoRequest): IPInfo {
        return generated<IPInfo, com.newnanmanager.generated.models.IPInfo> {
            api.getIPInfoWithHttpInfo(request.ip)
        }
    }

    fun banIP(request: BanIPRequest) {
        val body = toGenerated<com.newnanmanager.generated.models.BanIPRequest>(request)
        generatedVoid { api.banIPWithHttpInfo(body) }
    }

    fun unbanIP(request: UnbanIPRequest) {
        val body = toGenerated<com.newnanmanager.generated.models.UnbanIPRequest>(request)
        generatedVoid { api.unbanIPWithHttpInfo(body) }
    }

    fun listIPs(request: ListIPsRequest): ListIPsResponse {
        val level = request.minThreatLevel?.let { value ->
            com.newnanmanager.generated.apis.IPServiceApi.MinThreatLevelListIPs.values()
                .first { it.value.toInt() == value.value }
        }
        return generated<ListIPsResponse, com.newnanmanager.generated.models.ListIPsResponse> {
            api.listIPsWithHttpInfo(request.page, request.pageSize, request.bannedOnly, level, request.minRiskScore)
        }
    }

    fun getBannedIPs(request: ListIPsRequest): ListIPsResponse {
        val level = request.minThreatLevel?.let { value ->
            com.newnanmanager.generated.apis.IPServiceApi.MinThreatLevelGetBannedIPs.values()
                .first { it.value.toInt() == value.value }
        }
        return generated<ListIPsResponse, com.newnanmanager.generated.models.ListIPsResponse> {
            api.getBannedIPsWithHttpInfo(request.page, request.pageSize, request.bannedOnly, level, request.minRiskScore)
        }
    }

    fun getSuspiciousIPs(request: ListIPsRequest): ListIPsResponse {
        val level = request.minThreatLevel?.let { value ->
            com.newnanmanager.generated.apis.IPServiceApi.MinThreatLevelGetSuspiciousIPs.values()
                .first { it.value.toInt() == value.value }
        }
        return generated<ListIPsResponse, com.newnanmanager.generated.models.ListIPsResponse> {
            api.getSuspiciousIPsWithHttpInfo(request.page, request.pageSize, request.bannedOnly, level, request.minRiskScore)
        }
    }

    fun getHighRiskIPs(request: ListIPsRequest): ListIPsResponse {
        val level = request.minThreatLevel?.let { value ->
            com.newnanmanager.generated.apis.IPServiceApi.MinThreatLevelGetHighRiskIPs.values()
                .first { it.value.toInt() == value.value }
        }
        return generated<ListIPsResponse, com.newnanmanager.generated.models.ListIPsResponse> {
            api.getHighRiskIPsWithHttpInfo(request.page, request.pageSize, request.bannedOnly, level, request.minRiskScore)
        }
    }

    fun getIPStatistics(): IPStatistics {
        return generated<IPStatistics, com.newnanmanager.generated.models.IPStatistics> {
            api.getIPStatisticsWithHttpInfo()
        }
    }
}
